File: javascript和css学习/merge_sort2.js
//为了使数组原地排序，构造splice方法的参数，对数组进行先删后增。
function merge(left,right){
    var result = [],
        l = left.length,
        r = right.length,
        il = 0,
        ir = 0;
    while(il < l && ir < r){
        if(left[il] <= right[ir]){
            result.push(left[il++]);
        }else{
            result.push(right[ir++]);
        }
    }
    return result.concat(left.slice(il)).concat(right.slice(ir));
}

function mergeSort(a){
    var n = a.length;
    if(n < 2){
        return a;
    }
    var mid = Math.floor(n/2),
        left = a.slice(0,mid),
        right = a.slice(mid);
    var params = merge(mergeSort(left),mergeSort(right));
    params.unshift(0,n);
    a.splice.apply(a,params)
    return a;
}

File: merge_sort2.py
#python利用闭包实现就地排序
def merge_sort(a):
    n = len(a)
    if n<=1:
        return
    mid = n//2
    left = a[:mid]
    right = a[mid:]
    merge_sort(left)
    merge_sort(right)
    def merge(left,right):
        i = j = k = 0
        l = len(left)
        r = len(right)
        while j<l and k<r:
            if left[j] <= right[k]:
                a[i] = left[j]
                j+=1
            else:
                a[i] = right[k]
                k+=1
            i += 1
        while j<l:
            a[i] = left[j]
            i += 1
            j += 1
        while k<r:
            a[i] = right[k]
            i += 1
            k += 1
    merge(left,right)